forgevtt-module.js

```diff
@@ -1,2530 +1,2556 @@
+        }
+    }
+}
+/*
+// For testing
+ForgeVTT.HEARTBEAT_TIMER = 1 * 60 * 1000;
+ForgeVTT.ACTIVITY_CHECK_INTERVAL = 10 * 1000;
+ForgeVTT.ACTIVITY_UPDATE_INTERVAL = 60 * 1000;
+ForgeVTT.GAME_INACTIVE_THRESHOLD = 3 * 60 * 1000;
+ForgeVTT.OTHER_INACTIVE_THRESHOLD = 1 * 60 * 1000;
+ForgeVTT.HEARTBEAT_ACTIVE_IN_LAST_EVENTS = 10;
+ForgeVTT.IDLE_WARN_ADVANCE = 60 * 1000;
+*/
+
+class ForgeAPI_RateMonitor {
+    static reset(softReset = false) {
+        this.debug = false; // ForgeAPI_RateMonitor.debug = true to log trace on every call
+        this.timePeriod = 60 * 1000; // 1 minute; If the period changes, please review thresholds and APIRateMonitor messages
+        this.warningFrequency = 10; // Every 10 calls to the same endpoint after a threshold logs another warning
+        this.spikeWarningThreshold = 1000; // >= 1000 calls per minute indicates a spike
+        this.sustainedUsageMonitorThreshold = 6 * this.warningFrequency; // >= 60 calls per minute may indicate sustained usage
+        this.sustainedUsageWarningThreshold = 5; // Consecutive minutes that the monitor threshold has been hit
+        this.monitoring = true;
+        this.timeoutScheduled = false;
+        this.tracker = this.tracker || {}; // Each endpoint called will be a key in this object
+        if (softReset) {
+            // Soft reset keeps count of consecutive periods that the warning threshold was reached per endpoint
+            Object.keys(this.tracker).forEach((endpoint) => {
+                // { calls: number, consecutive: number }
+                if (this.tracker[endpoint].calls >= this.sustainedUsageMonitorThreshold) {
+                    this.tracker[endpoint].consecutive++; // Increment the consecutive counter
+                    this.tracker[endpoint].calls = 0; // Reset the call counter
+                } else {
+                    delete this.tracker[endpoint]; // Not consecutively above threshold. Stop tracking this endpoint
+                }
+            });
+        } else {
+            this.tracker = {};
+        }
+    }
+
+    static monitor(endpoint) {
+        try {
+            if (!this.monitoring) {
+                this.reset();
+            }
+            // Initialize the tracker for this endpoint if not already present
+            if (!this.tracker[endpoint]) {
+                this.tracker[endpoint] = { calls: 0, consecutive: 0 };
+            }
+            // Increment the counter for this endpoint
+            this.tracker[endpoint].calls++;
+            if (this.debug) {
+                this.logTrace(endpoint);
+            }
+            // Usage Spike: Warn per 10 calls if >= 1000 calls to the same endpoint in the current minute
+            if (
+                this.tracker[endpoint].calls >= this.spikeWarningThreshold &&
+                this.tracker[endpoint].calls % this.warningFrequency === 0
+            ) {
+                const warning = game.i18n.format("THEFORGE.APIRateMonitorSpikeWarning", {
```

Comment:
I'm worried this will be excessively spammy. If they've reached >=1000 calls in a minute, making a UI notification every 10 calls is probably going to be way too much. That user with 1200 calls in a minute would have gotten 20 notifications.

Maybe once when they hit 1000 and again every 100 calls (warningFrequency * 10) after that in the same minute?

Reply:
I do agree with you, but there is one complication.

In the case of sustained usage, which usually indicates that server side rate limiting is in effect or that the request is repetitive but not recursive, the rate will hover at under 100 requests a second (observed in testing, at least). In that case, we may never reach the frequency of 100.

The main reason this is a concern is that if the user has seen the warning and has opened their browser, we want them to be able to see the trace log without having to re-trigger the warning from scratch. My concern with 100 was that that would be too high, and the user may feel confused with why there's nothing happening after being told to open the console.

I think the solution might be a multiplier of 10 for the frequency check if it's a spike. That way, we issue spike warnings every 100, but sustained use warnings every 10. That's more in line with the frequencies I observed in testing, and will result in fewer nuisance alerts for users.

Reply:
Yes, agreed. Keep `warningFrequency` as it is, I think it has a good value. I was referring just to the `if` condition on lines 1500 and 1501 (guess I picked the wrong line to comment on for it to be visible in the conversation tab), which I think should be `this.tracker[endpoint].calls >= this.spikeWarningThreshold && this.tracker[endpoint].calls % (this.warningFrequency * 10) === 0` or similar to avoid spam.

Reply:
Ack 👍